File: airbyte-integrations/bases/base-java-s3/src/main/java/io/airbyte/integrations/destination/s3/S3DestinationConfig.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.s3;

import static io.airbyte.integrations.destination.s3.constant.S3Constants.ACCESS_KEY_ID;
import static io.airbyte.integrations.destination.s3.constant.S3Constants.ACCOUNT_ID;
import static io.airbyte.integrations.destination.s3.constant.S3Constants.FILE_NAME_PATTERN;
import static io.airbyte.integrations.destination.s3.constant.S3Constants.SECRET_ACCESS_KEY;
import static io.airbyte.integrations.destination.s3.constant.S3Constants.S_3_BUCKET_NAME;
import static io.airbyte.integrations.destination.s3.constant.S3Constants.S_3_BUCKET_PATH;
import static io.airbyte.integrations.destination.s3.constant.S3Constants.S_3_BUCKET_REGION;
import static io.airbyte.integrations.destination.s3.constant.S3Constants.S_3_ENDPOINT;
import static io.airbyte.integrations.destination.s3.constant.S3Constants.S_3_PATH_FORMAT;

import com.amazonaws.ClientConfiguration;
import com.amazonaws.Protocol;
import com.amazonaws.auth.AWSCredentialsProvider;
import com.amazonaws.client.builder.AwsClientBuilder;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.fasterxml.jackson.databind.JsonNode;
import io.airbyte.integrations.destination.s3.credential.S3AWSDefaultProfileCredentialConfig;
import io.airbyte.integrations.destination.s3.credential.S3AccessKeyCredentialConfig;
import io.airbyte.integrations.destination.s3.credential.S3CredentialConfig;
import io.airbyte.integrations.destination.s3.credential.S3CredentialType;
import java.util.Objects;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * An S3 configuration. Typical usage sets at most one of {@code bucketPath} (necessary for more
 * delicate data syncing to S3)
 */
public class S3DestinationConfig {

  private static final Logger LOGGER = LoggerFactory.getLogger(S3DestinationConfig.class);
  private static final String R2_INSTANCE_URL = "https://%s.r2.cloudflarestorage.com";

  private final String endpoint;
  private final String bucketName;
  private final String bucketPath;
  private final String bucketRegion;
  private final String pathFormat;
  private final S3CredentialConfig credentialConfig;
  private final S3FormatConfig formatConfig;
  private String fileNamePattern;

  private final Object lock = new Object();
  private AmazonS3 s3Client;

  private boolean checkIntegrity = true;

  private int uploadThreadsCount = S3StorageOperations.DEFAULT_UPLOAD_THREADS;

  public S3DestinationConfig(final String endpoint,
                             final String bucketName,
                             final String bucketPath,
                             final String bucketRegion,
                             final String pathFormat,
                             final S3CredentialConfig credentialConfig,
                             final S3FormatConfig formatConfig,
                             final AmazonS3 s3Client) {
    this.endpoint = endpoint;
    this.bucketName = bucketName;
    this.bucketPath = bucketPath;
    this.bucketRegion = bucketRegion;
    this.pathFormat = pathFormat;
    this.credentialConfig = credentialConfig;
    this.formatConfig = formatConfig;
    this.s3Client = s3Client;
  }

  public S3DestinationConfig(final String endpoint,
                             final String bucketName,
                             final String bucketPath,
                             final String bucketRegion,
                             final String pathFormat,
                             final S3CredentialConfig credentialConfig,
                             final S3FormatConfig formatConfig,
                             final AmazonS3 s3Client,
                             final String fileNamePattern,
                             final boolean checkIntegrity,
                             final int uploadThreadsCount) {
    this.endpoint = endpoint;
    this.bucketName = bucketName;
    this.bucketPath = bucketPath;
    this.bucketRegion = bucketRegion;
    this.pathFormat = pathFormat;
    this.credentialConfig = credentialConfig;
    this.formatConfig = formatConfig;
    this.s3Client = s3Client;
    this.fileNamePattern = fileNamePattern;
    this.checkIntegrity = checkIntegrity;
    this.uploadThreadsCount = uploadThreadsCount;
  }

  public static Builder create(final String bucketName, final String bucketPath, final String bucketRegion) {
    return new Builder(bucketName, bucketPath, bucketRegion);
  }

  public static Builder create(final S3DestinationConfig config) {
    return new Builder(config.getBucketName(), config.getBucketPath(), config.getBucketRegion())
        .withEndpoint(config.getEndpoint())
        .withCredentialConfig(config.getS3CredentialConfig())
        .withFormatConfig(config.getFormatConfig());
  }

  public static S3DestinationConfig getS3DestinationConfig(@Nonnull final JsonNode config) {
    return getS3DestinationConfig(config, StorageProvider.AWS_S3);
  }

  public static S3DestinationConfig getS3DestinationConfig(@Nonnull final JsonNode config, @Nonnull final StorageProvider storageProvider) {
    Builder builder = create(
        getProperty(config, S_3_BUCKET_NAME),
        "",
        getProperty(config, S_3_BUCKET_REGION));

    if (config.has(S_3_BUCKET_PATH)) {
      builder = builder.withBucketPath(config.get(S_3_BUCKET_PATH).asText());
    }

    if (config.has(FILE_NAME_PATTERN)) {
      builder = builder.withFileNamePattern(config.get(FILE_NAME_PATTERN).asText());
    }

    if (config.has(S_3_PATH_FORMAT)) {
      builder = builder.withPathFormat(config.get(S_3_PATH_FORMAT).asText());
    }

    switch (storageProvider) {
      case CF_R2 -> {
        if (config.has(ACCOUNT_ID)) {
          final String endpoint = String.format(R2_INSTANCE_URL, getProperty(config, ACCOUNT_ID));
          builder = builder.withEndpoint(endpoint);
        }
        builder = builder.withCheckIntegrity(false)
            // https://developers.cloudflare.com/r2/platform/s3-compatibility/api/#implemented-object-level-operations
            // 3 or less
            .withUploadThreadsCount(S3StorageOperations.R2_UPLOAD_THREADS);
      }
      default -> {
        if (config.has(S_3_ENDPOINT)) {
          builder = builder.withEndpoint(config.get(S_3_ENDPOINT).asText());
        }
      }
    }

    final S3CredentialConfig credentialConfig;
    if (config.has(ACCESS_KEY_ID)) {
      credentialConfig = new S3AccessKeyCredentialConfig(getProperty(config, ACCESS_KEY_ID), getProperty(config, SECRET_ACCESS_KEY));
    } else {
      credentialConfig = new S3AWSDefaultProfileCredentialConfig();
    }
    builder = builder.withCredentialConfig(credentialConfig);

    // In the "normal" S3 destination, this is never null. However, the Redshift and Snowflake copy
    // destinations don't set a Format config.
    if (config.has("format")) {
      builder = builder.withFormatConfig(S3FormatConfigs.getS3FormatConfig(config));
    }

    return builder.get();
  }

  @Nullable
  private static String getProperty(@Nonnull final JsonNode config, @Nonnull final String key) {
    final JsonNode node = config.get(key);
    if (node == null) {
      return null;
    }
    return node.asText();
  }

  public String getEndpoint() {
    return endpoint;
  }

  public String getBucketName() {
    return bucketName;
  }

  public String getBucketPath() {
    return bucketPath;
  }

  public String getPathFormat() {
    return pathFormat;
  }

  public String getBucketRegion() {
    return bucketRegion;
  }

  public String getFileNamePattern() {
    return fileNamePattern;
  }

  public S3CredentialConfig getS3CredentialConfig() {
    return credentialConfig;
  }

  public S3FormatConfig getFormatConfig() {
    return formatConfig;
  }

  public boolean isCheckIntegrity() {
    return checkIntegrity;
  }

  public int getUploadThreadsCount() {
    return uploadThreadsCount;
  }

  public AmazonS3 getS3Client() {
    synchronized (lock) {
      if (s3Client == null) {
        return resetS3Client();
      }
      return s3Client;
    }
  }

  AmazonS3 resetS3Client() {
    synchronized (lock) {
      if (s3Client != null) {
        s3Client.shutdown();
      }
      s3Client = createS3Client();
      return s3Client;
    }
  }

  protected AmazonS3 createS3Client() {
    LOGGER.info("Creating S3 client...");

    final AWSCredentialsProvider credentialsProvider = credentialConfig.getS3CredentialsProvider();
    final S3CredentialType credentialType = credentialConfig.getCredentialType();

    if (S3CredentialType.DEFAULT_PROFILE == credentialType) {
      return AmazonS3ClientBuilder.standard()
          .withRegion(bucketRegion)
          .withCredentials(credentialsProvider)
          .build();
    }

    if (null == endpoint || endpoint.isEmpty()) {
      return AmazonS3ClientBuilder.standard()
          .withCredentials(credentialsProvider)
          .withRegion(bucketRegion)
          .build();
    }

    final ClientConfiguration clientConfiguration = new ClientConfiguration().withProtocol(Protocol.HTTPS);
    clientConfiguration.setSignerOverride("AWSS3V4SignerType");

    return AmazonS3ClientBuilder.standard()
        .withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, bucketRegion))
        .withPathStyleAccessEnabled(true)
        .withClientConfiguration(clientConfiguration)
        .withCredentials(credentialsProvider)
        .build();
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    final S3DestinationConfig that = (S3DestinationConfig) o;
    return Objects.equals(endpoint, that.endpoint) && Objects.equals(bucketName, that.bucketName) && Objects.equals(
        bucketPath, that.bucketPath) && Objects.equals(bucketRegion, that.bucketRegion)
        && Objects.equals(credentialConfig, that.credentialConfig)
        && Objects.equals(formatConfig, that.formatConfig);
  }

  @Override
  public int hashCode() {
    return Objects.hash(endpoint, bucketName, bucketPath, bucketRegion, credentialConfig, formatConfig);
  }

  public static class Builder {

    private String endpoint = "";
    private String pathFormat = S3DestinationConstants.DEFAULT_PATH_FORMAT;

    private String bucketName;
    private String bucketPath;
    private String bucketRegion;
    private S3CredentialConfig credentialConfig;
    private S3FormatConfig formatConfig;
    private AmazonS3 s3Client;
    private String fileNamePattern;

    private boolean checkIntegrity = true;

    private int uploadThreadsCount = S3StorageOperations.DEFAULT_UPLOAD_THREADS;

    protected Builder(final String bucketName, final String bucketPath, final String bucketRegion) {
      this.bucketName = bucketName;
      this.bucketPath = bucketPath;
      this.bucketRegion = bucketRegion;
    }

    public Builder withBucketName(final String bucketName) {
      this.bucketName = bucketName;
      return this;
    }

    public Builder withFileNamePattern(final String fileNamePattern) {
      this.fileNamePattern = fileNamePattern;
      return this;
    }

    public Builder withBucketPath(final String bucketPath) {
      this.bucketPath = bucketPath;
      return this;
    }

    public Builder withBucketRegion(final String bucketRegion) {
      this.bucketRegion = bucketRegion;
      return this;
    }

    public Builder withPathFormat(final String pathFormat) {
      this.pathFormat = pathFormat;
      return this;
    }

    public Builder withEndpoint(final String endpoint) {
      this.endpoint = endpoint;
      return this;
    }

    public Builder withFormatConfig(final S3FormatConfig formatConfig) {
      this.formatConfig = formatConfig;
      return this;
    }

    public Builder withAccessKeyCredential(final String accessKeyId, final String secretAccessKey) {
      this.credentialConfig = new S3AccessKeyCredentialConfig(accessKeyId, secretAccessKey);
      return this;
    }

    public Builder withCredentialConfig(final S3CredentialConfig credentialConfig) {
      this.credentialConfig = credentialConfig;
      return this;
    }

    public Builder withS3Client(final AmazonS3 s3Client) {
      this.s3Client = s3Client;
      return this;
    }

    public Builder withCheckIntegrity(final boolean checkIntegrity) {
      this.checkIntegrity = checkIntegrity;
      return this;
    }

    public Builder withUploadThreadsCount(final int uploadThreadsCount) {
      this.uploadThreadsCount = uploadThreadsCount;
      return this;
    }

    public S3DestinationConfig get() {
      return new S3DestinationConfig(
          endpoint,
          bucketName,
          bucketPath,
          bucketRegion,
          pathFormat,
          credentialConfig,
          formatConfig,
          s3Client,
          fileNamePattern,
          checkIntegrity,
          uploadThreadsCount);
    }

  }

}

File: airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/S3DestinationConfig.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.s3

import com.amazonaws.ClientConfiguration
import com.amazonaws.Protocol
import com.amazonaws.client.builder.AwsClientBuilder
import com.amazonaws.retry.RetryMode
import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.AmazonS3ClientBuilder
import com.fasterxml.jackson.databind.JsonNode
import io.airbyte.cdk.integrations.destination.s3.constant.S3Constants
import io.airbyte.cdk.integrations.destination.s3.credential.*
import io.github.oshai.kotlinlogging.KotlinLogging
import java.util.*
import javax.annotation.Nonnull

private val LOGGER = KotlinLogging.logger {}
/**
 * An S3 configuration. Typical usage sets at most one of `bucketPath` (necessary for more delicate
 * data syncing to S3)
 */
open class S3DestinationConfig {
    val endpoint: String?
    val bucketName: String?
    val bucketPath: String?
    val bucketRegion: String?
    val pathFormat: String?
    val s3CredentialConfig: S3CredentialConfig?
    val formatConfig: UploadFormatConfig?
    var fileNamePattern: String? = null
        private set
    var environment: Map<String, String>

    private val lock = Any()
    private var s3Client: AmazonS3?
    fun getS3Client(): AmazonS3 {
        synchronized(lock) {
            s3Client?.let {
                return it
            }
            return resetS3Client()
        }
    }

    var isCheckIntegrity: Boolean = true
        private set

    var uploadThreadsCount: Int = S3StorageOperations.DEFAULT_UPLOAD_THREADS
        private set

    constructor(
        endpoint: String?,
        bucketName: String,
        bucketPath: String?,
        bucketRegion: String?,
        pathFormat: String?,
        credentialConfig: S3CredentialConfig?,
        formatConfig: UploadFormatConfig?,
        s3Client: AmazonS3
    ) {
        this.endpoint = endpoint
        this.bucketName = bucketName
        this.bucketPath = bucketPath
        this.bucketRegion = bucketRegion
        this.pathFormat = pathFormat
        this.s3CredentialConfig = credentialConfig
        this.formatConfig = formatConfig
        this.s3Client = s3Client
        this.environment = System.getenv()
    }

    constructor(
        endpoint: String?,
        bucketName: String?,
        bucketPath: String?,
        bucketRegion: String?,
        pathFormat: String?,
        credentialConfig: S3CredentialConfig?,
        formatConfig: UploadFormatConfig?,
        s3Client: AmazonS3?,
        fileNamePattern: String?,
        checkIntegrity: Boolean,
        uploadThreadsCount: Int,
        environment: Map<String, String> = System.getenv()
    ) {
        this.endpoint = endpoint
        this.bucketName = bucketName
        this.bucketPath = bucketPath
        this.bucketRegion = bucketRegion
        this.pathFormat = pathFormat
        this.s3CredentialConfig = credentialConfig
        this.formatConfig = formatConfig
        this.s3Client = s3Client
        this.fileNamePattern = fileNamePattern
        this.isCheckIntegrity = checkIntegrity
        this.uploadThreadsCount = uploadThreadsCount
        this.environment = environment
    }

    fun resetS3Client(): AmazonS3 {
        synchronized(lock) {
            s3Client?.shutdown()
            val s3Client = createS3Client()
            this.s3Client = s3Client
            return s3Client
        }
    }

    protected open fun createS3Client(): AmazonS3 {
        LOGGER.info { "Creating S3 client..." }

        val credentialsProvider = s3CredentialConfig!!.s3CredentialsProvider
        val credentialType = s3CredentialConfig.credentialType

        val clientBuilder = AmazonS3ClientBuilder.standard().withCredentials(credentialsProvider)
        when (credentialType) {
            S3CredentialType.DEFAULT_PROFILE,
            S3CredentialType.ASSUME_ROLE ->
                clientBuilder
                    .withRegion(bucketRegion)
                    // the SDK defaults to RetryMode.LEGACY
                    // (https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html)
                    // this _can_ be configured via environment variable, but it seems more reliable
                    // to
                    // configure it
                    // programmatically
                    .withClientConfiguration(
                        ClientConfiguration().withRetryMode(RetryMode.STANDARD)
                    )
            S3CredentialType.ACCESS_KEY -> {
                if (null == endpoint || endpoint.isEmpty()) {
                    clientBuilder.withRegion(bucketRegion)
                } else {
                    val clientConfiguration = ClientConfiguration().withProtocol(Protocol.HTTPS)
                    clientConfiguration.signerOverride = "AWSS3V4SignerType"

                    clientBuilder
                        .withEndpointConfiguration(
                            AwsClientBuilder.EndpointConfiguration(endpoint, bucketRegion)
                        )
                        .withPathStyleAccessEnabled(true)
                        .withClientConfiguration(clientConfiguration)
                }
            }
        }
        return clientBuilder.build()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as S3DestinationConfig
        return endpoint == that.endpoint &&
            bucketName == that.bucketName &&
            bucketPath == that.bucketPath &&
            bucketRegion == that.bucketRegion &&
            s3CredentialConfig == that.s3CredentialConfig &&
            formatConfig == that.formatConfig
    }

    override fun hashCode(): Int {
        return Objects.hash(
            endpoint,
            bucketName,
            bucketPath,
            bucketRegion,
            s3CredentialConfig,
            formatConfig
        )
    }

    class Builder(
        private var bucketName: String?,
        private var bucketPath: String,
        private var bucketRegion: String?
    ) {
        private var endpoint: String? = ""
        private var pathFormat = S3DestinationConstants.DEFAULT_PATH_FORMAT

        private lateinit var credentialConfig: S3CredentialConfig
        private var formatConfig: UploadFormatConfig? = null
        private var s3Client: AmazonS3? = null
        private var fileNamePattern: String? = null

        private var checkIntegrity = true

        private var uploadThreadsCount = S3StorageOperations.DEFAULT_UPLOAD_THREADS
        private var environment: Map<String, String> = System.getenv()

        fun withBucketName(bucketName: String): Builder {
            this.bucketName = bucketName
            return this
        }

        fun withFileNamePattern(fileNamePattern: String?): Builder {
            this.fileNamePattern = fileNamePattern
            return this
        }

        fun withBucketPath(bucketPath: String): Builder {
            this.bucketPath = bucketPath
            return this
        }

        fun withBucketRegion(bucketRegion: String?): Builder {
            this.bucketRegion = bucketRegion
            return this
        }

        fun withPathFormat(pathFormat: String): Builder {
            this.pathFormat = pathFormat
            return this
        }

        fun withEndpoint(endpoint: String?): Builder {
            this.endpoint = endpoint
            return this
        }

        fun withFormatConfig(formatConfig: UploadFormatConfig?): Builder {
            this.formatConfig = formatConfig
            return this
        }

        fun withAccessKeyCredential(accessKeyId: String?, secretAccessKey: String?): Builder {
            this.credentialConfig = S3AccessKeyCredentialConfig(accessKeyId, secretAccessKey)
            return this
        }

        fun withCredentialConfig(credentialConfig: S3CredentialConfig): Builder {
            this.credentialConfig = credentialConfig
            return this
        }

        fun withS3Client(s3Client: AmazonS3): Builder {
            this.s3Client = s3Client
            return this
        }

        fun withCheckIntegrity(checkIntegrity: Boolean): Builder {
            this.checkIntegrity = checkIntegrity
            return this
        }

        fun withUploadThreadsCount(uploadThreadsCount: Int): Builder {
            this.uploadThreadsCount = uploadThreadsCount
            return this
        }

        fun withEnvironment(environment: Map<String, String>): Builder {
            this.environment = environment
            return this
        }

        fun get(): S3DestinationConfig {
            return S3DestinationConfig(
                endpoint,
                bucketName,
                bucketPath,
                bucketRegion,
                pathFormat,
                credentialConfig,
                formatConfig,
                s3Client,
                fileNamePattern,
                checkIntegrity,
                uploadThreadsCount,
                environment
            )
        }
    }

    companion object {

        private const val R2_INSTANCE_URL = "https://%s.r2.cloudflarestorage.com"

        @JvmStatic
        fun create(bucketName: String?, bucketPath: String, bucketRegion: String?): Builder {
            return Builder(bucketName, bucketPath, bucketRegion)
        }

        @JvmStatic
        fun create(config: S3DestinationConfig): Builder {
            return Builder(config.bucketName, config.bucketPath!!, config.bucketRegion)
                .withEndpoint(config.endpoint)
                .withCredentialConfig(config.s3CredentialConfig!!)
                .withFormatConfig(config.formatConfig)
        }

        @JvmStatic
        fun getS3DestinationConfig(
            @Nonnull config: JsonNode,
            environment: Map<String, String> = System.getenv()
        ): S3DestinationConfig {
            return getS3DestinationConfig(config, StorageProvider.AWS_S3, environment)
        }

        @JvmStatic
        fun getS3DestinationConfig(
            @Nonnull config: JsonNode,
            @Nonnull storageProvider: StorageProvider = StorageProvider.AWS_S3,
            environment: Map<String, String> = System.getenv()
        ): S3DestinationConfig {
            var builder =
                create(
                    getProperty(config, S3Constants.S_3_BUCKET_NAME),
                    "",
                    getProperty(config, S3Constants.S_3_BUCKET_REGION)
                )

            if (config.has(S3Constants.S_3_BUCKET_PATH)) {
                builder = builder.withBucketPath(config[S3Constants.S_3_BUCKET_PATH].asText())
            }

            if (config.has(S3Constants.FILE_NAME_PATTERN)) {
                builder =
                    builder.withFileNamePattern(config[S3Constants.FILE_NAME_PATTERN].asText())
            }

            if (config.has(S3Constants.S_3_PATH_FORMAT)) {
                builder = builder.withPathFormat(config[S3Constants.S_3_PATH_FORMAT].asText())
            }

            when (storageProvider) {
                StorageProvider.CF_R2 -> {
                    if (config.has(S3Constants.ACCOUNT_ID)) {
                        val endpoint =
                            String.format(
                                R2_INSTANCE_URL,
                                getProperty(config, S3Constants.ACCOUNT_ID)
                            )
                        builder = builder.withEndpoint(endpoint)
                    }
                    builder =
                        builder
                            .withCheckIntegrity(
                                false
                            ) // https://developers.cloudflare.com/r2/platform/s3-compatibility/api/#implemented-object-level-operations
                            // 3 or less
                            .withUploadThreadsCount(S3StorageOperations.R2_UPLOAD_THREADS)
                }
                else -> {
                    if (config.has(S3Constants.S_3_ENDPOINT)) {
                        builder = builder.withEndpoint(config[S3Constants.S_3_ENDPOINT].asText())
                    }
                }
            }
            val credentialConfig =
                if (config.has(S3Constants.ACCESS_KEY_ID)) {
                    S3AccessKeyCredentialConfig(
                        getProperty(config, S3Constants.ACCESS_KEY_ID),
                        getProperty(config, S3Constants.SECRET_ACCESS_KEY)
                    )
                } else if (config.has(S3Constants.ROLE_ARN)) {
                    S3AssumeRoleCredentialConfig(
                        getProperty(config, S3Constants.ROLE_ARN)!!,
                        environment
                    )
                } else {
                    S3AWSDefaultProfileCredentialConfig()
                }
            builder = builder.withCredentialConfig(credentialConfig)

            // In the "normal" S3 destination, this is never null. However, the Redshift and
            // Snowflake copy
            // destinations don't set a Format config.
            if (config.has("format")) {
                builder =
                    builder.withFormatConfig(
                        UploadFormatConfigFactory.getUploadFormatConfig(config)
                    )
            }
            builder.withEnvironment(environment)
            return builder.get()
        }

        private fun getProperty(config: JsonNode, @Nonnull key: String): String? {
            val node: JsonNode? = config.get(key)
            return node?.asText()
        }
    }
}
